File: algorithms/c#/QuickSort.cs
namespace Algorithms
{
	public static class QuickSort
	{
		private static int Partition(int[] array, int start, int end)
		{
			int pivot = array[start];
			int median = start;

			for(int i = start + 1; i <= end; i++)
			{
				if(array[i] < pivot)
				{
					median++;
					int a = array[i];
					int b = array[median];
					array[i] = b;
					array[median] = a;
				}
			}

			int x = array[start];
			int y = array[median];
			array[start] = y;
			array[median] = x;

			return median;
		}

		public static void Sort(int[] array, int start, int end)
		{
			if(start < end)
			{
				int pivot = Partition(array, start, end);

				Sort(array, start, pivot - 1);
				Sort(array, pivot + 1, end);
			}
		}
	}
}

File: algorithms/go/quick_sort.go
package algorithms

func Partition(slice []int, start int, end int) int {
	median := start
	pivot := slice[start]

	for i := start; i < end; i++ {
		median++

		if slice[i] < pivot {
			temp1 := slice[i]
			temp2 := slice[median]

			slice[i] = temp2
			slice[median] = temp1
		}
	}

	temp1 := slice[start]
	temp2 := slice[median]

	slice[start] = temp2
	slice[median] = temp1

	return median
}

func QuickSort(slice []int, start int, end int) {
	if start < end {
		pivot := Partition(slice, start, end)

		QuickSort(slice, start, pivot - 1)
		QuickSort(slice, pivot + 1, end)
	}
}